### PR TITLE
Optimize switch codegen and improve docs

### LINKED ISSUE #665: cranelift_frontend::Switch produces non optimal code
Diff between current and wanted:

```diff
 function u0:0() system_v {
 ; symbol _ZN21mini_core_hello_world4main17he1655d58f9d91204E
 ; instance Instance { def: Item(DefId(0/0:29 ~ mini_core_hello_world[317d]::main[0])), substs: [] }
 ; sig ([]; variadic: false)->()
 
 ; ssa {_1: NOT_SSA, _2: (empty), _0: NOT_SSA}
 ; msg   loc.idx    param    pass mode            ssa flags  ty
 ; ret    _0      -          NoPass               NOT_SSA    ()
 
     ss0 = explicit_slot 1 ; _1: Ordering size=1 align=1,8
     jt0 = jump_table [ebb2, ebb4]
 
 ebb0:
         v0 = stack_addr.i64 ss0
         jump ebb1
 
 ebb1:
 ; discriminant(_1) = 1
         v1 = iconst.i8 0
         store v1, v0
 ; _2 = discriminant(_1)
         v2 = load.i8 v0
         v3 = sextend.i64 v2
 ; 
 ; switchInt(move _2)
         v4 = icmp_imm eq v3, -1
         brnz v4, ebb2
-        v5 = icmp_imm uge v3, 0
-        brnz v5, ebb5
+        brz v3, ebb5
         jump ebb3

 ebb5:
-        v6 = iadd_imm.i64 v3, 0
-        br_table v6, ebb3, jt0
+        br_table v3, ebb3, jt0
 
 ebb2:
 ; 
 ; return
         return
 
 ebb3:
 ; 
 ; unreachable
         trap user65535
 
 ebb4:
 ; 
 ; goto
         jump ebb4
 }
```

An aditional optimization for this case would be #606:

```diff
-        brz v3, ebb5
-        jump ebb3
-
-ebb5:
+        brnz v3, ebb3
```

### PR DESCRIPTION
**Pull request by bjorn3**
Fixes #665

### CODE REVIEW

File: cranelift-frontend/src/switch.rs [hunk 1]
```diff
@@ -129,21 +169,38 @@ impl Switch {
     ) {
         for ContiguousCaseRange { first_index, ebbs } in contiguous_case_ranges.into_iter().rev() {
             if ebbs.len() == 1 {
-                let is_good_val = bx.ins().icmp_imm(IntCC::Equal, val, first_index as i64);
-                bx.ins().brnz(is_good_val, ebbs[0], &[]);
+                if first_index == 0 {
```
Comment (bnjbvr) [inline cranelift-frontend/src/switch.rs]: (light suggestion) There might be a way to reduce the number of code indents here by using `match (ebbs.len(), first_index)` with 4 cases.

File: cranelift-frontend/src/switch.rs [hunk 2]
```diff
@@ -129,21 +169,38 @@ impl Switch {
...
+                }
             } else {
                 let jt_ebb = bx.create_ebb();
-                let is_good_val =
-                    bx.ins()
-                        .icmp_imm(IntCC::UnsignedGreaterThanOrEqual, val, first_index as i64);
-                bx.ins().brnz(is_good_val, jt_ebb, &[]);
-                cases_and_jt_ebbs.push((first_index, jt_ebb, ebbs));
+                if first_index == 0 {
+                    // if first_index is 0, then icmp_imm uge val, first_index is trivially true
```
Comment (bnjbvr) [inline cranelift-frontend/src/switch.rs]: nit: can you put backquotes around the cranelift instruction, and parenthesis within it, to make it more readable? (and add a dot at the end)

File: cranelift-frontend/src/switch.rs [hunk 3]
```diff
@@ -9,6 +9,36 @@ type EntryIndex = u64;
...
 /// Unlike with `br_table`, `Switch` cases may be sparse or non-0-based.
 /// They emit efficient code using branches, jump tables, or a combination of both.
+///
+/// # Example
+///
+/// ```rust
+/// # use cranelift_codegen::ir::types::*;
+/// # use cranelift_codegen::ir::{ExternalName, Function, Signature};
+/// # use cranelift_codegen::isa::CallConv;
+/// # use cranelift_frontend::{FunctionBuilder, FunctionBuilderContext, Switch};
```
Comment (bnjbvr) [inline cranelift-frontend/src/switch.rs]: CI broke, saying you need `use cranelift_codegen::ir::builder::InstBuilder;` here too to be able to use `builder.ins().iconst` below.

File: cranelift-frontend/src/switch.rs (hunk 2 above)
Comment (bjorn3) [reply cranelift-frontend/src/switch.rs]: > and parenthesis within it

Where?

Comment (bnjbvr) [reply cranelift-frontend/src/switch.rs]: I thought around the instructions, but the backquotes make it clear enough already, so this is fine as is. Thanks.

**Pull request merged**